README.md
# JavaScript package.json License Checker

[![npm Version][npm-image]][npm-url]
[![CI][actions-image]][actions-url]
[![Dependency Status][david-image]][david-url]
[![Known Vulnerabilities][snyk-image]][snyk-url]
[![Code Style: Google][gts-image]][gts-url]

**This is not an official Google product.**

This is a tool for checking the license of JavaScript projects. It scans the
`package.json` file to check its license and recursively checks all of its
dependencies.

**DISCLAIMER: This tool is NOT a replacement for legal advice or due
diligence for your project's license validity. We recommend you consult a
lawyer if you want legal advice.**

## Installation

```shell
npm install [--save-dev] js-green-licenses
```

If you want to install globally,

```shell
npm install -g js-green-licenses
```

## CLI

```
usage: jsgl [-h] [-v] [--local <directory>] [--pr <github PR>]
            [--dev] [--verbose] [<package or package@version>]

License checker for npm modules

Positional arguments:
  <package or package@version>
                        Package name to check license for. Can include
                        version spec after @. E.g. foo@^1.2.3. Otherwise
                        latest.

Optional arguments:
  -h, --help            Show this help message and exit.
  -v, --version         Show program's version number and exit.
  --local <directory>, -l <directory>
                        Check a local directory instead of public npm.
  --pr <github PR>      Check a github pull request. Must be
                        <owner>/<repo>/pull/<id>
  --dev                 Also check devDependencies.
  --verbose             Verbose error outputs.
```

This tool checks licenses for 1) an already published npm package, 2) a local
directory, or 3) a GitHub pull request. For checking an npm package, you can
just pass the package name (optionally together with the version) as the
argument. To check a local directory, you should pass the `--local
path/to/repo` argument. To check for a GitHub PR, you should pass the `--pr
<owner>/<repo>/pull/<id>` argument.

If the tool finds any non-green licenses in the given package or in its
dependencies, they will be printed out together with the detailed
information.

If you pass `--dev`, the `devDependencies` will be checked as well as the
`dependencies`.

`jsgl` also checks sub-packages for `--local` and `--pr` flags when it
detects that the repository is a monorepo. It assumes a certain directory
structure for detecting whether a repository is a monorepo: the top-level
directory should have the `packages` directory in it and sub-packages must
exist under that directory. In that case, all the `package.json` files are
found from sub-packages and `jsgl` checks all of them.

For example, when a directory `foo` is like this:
```
foo
 |
 +-- packages
 |    |
 |    +-- bar
 |    |    |
 |    |    +-- package.json
 |    |    |
 |    |    +-- ...
 |    |
 |    +-- baz
 |         |
 |         +-- package.json
 |         |
 |         +-- ...
 |
 +-- package.json
 |
 +-- ...
```
, `jsgl` checks all of `foo/package.json`, `foo/packages/bar/package.json`,
and `foo/packages/baz/package.json`.

## Configurations

You can customize how `jsgl` works with the configuration file, named
`js-green-licenses.json`. For example, you can specify the license list that
you would like to consider green. The license IDs must be listed in the
`greenLicenses` section of the configuration file. In that case, `jsgl` will
use that custom list instead of its default list.

The default green license list is:
```javascript
const DEFAULT_GREEN_LICENSES = [
  '0BSD',         'AFL-2.1',      'AFL-3.0',      'APSL-2.0',     'Apache-1.1',
  'Apache-2.0',   'Artistic-1.0', 'Artistic-2.0', 'BSD-2-Clause', 'BSD-3-Clause',
  'BSL-1.0',      'CC-BY-1.0',    'CC-BY-2.0',    'CC-BY-2.5',    'CC-BY-3.0',
  'CC-BY-4.0',    'CC0-1.0',      'CDDL-1.0',     'CDDL-1.1',     'CPL-1.0',
  'EPL-1.0',      'FTL',          'IPL-1.0',      'ISC',          'LGPL-2.0',
  'LGPL-2.1',     'LGPL-3.0',     'LPL-1.02',     'MIT',          'MPL-1.0',
  'MPL-1.1',      'MPL-2.0',      'MS-PL',        'NCSA',         'OpenSSL',
  'PHP-3.0',      'Ruby',         'Unlicense',    'W3C',          'Xnet',
  'ZPL-2.0',      'Zend-2.0',     'Zlib',         'libtiff',
];
```

You can also allowlist some npm packages and they will be considered "green"
even when they have non-green licenses or no licenses. It's useful when
`jsgl` is unable to verify the validness of a certain package's license for
some reason. For example, when a package doesn't specify its license in its
`package.json` but has a separate `LICENSE` file, `jsgl` can't verify that.
You can allowlist that package to make `jsgl` not complain about that
package.

A typical configuration file looks like this:
```javascript
{
  "greenLicenses": [
    // Custom green licenses.
    "Apache-2.0",
    "MIT",
    "BSD-3-Clause",
    ...
  ],
  "packageAllowlist": [
    /* packages considered ok */
    "foo",
    "bar",  // inline comment
    "package-with-no-license",
    "package-with-okish-license",
    ...
  ]
}
```

The `greenLicenses` section is for the custom license list and the
`packageAllowlist` section is for the package allowlist.

Note that comments are allowed in `js-green-licenses.json`.

The configuration file must be located in the top-level directory of a
repository for `--local` and `--pr`. When checking remote npm packages,
`jsgl` tries to locate the configuration file in the current local directory
from which `jsgl` is invoked.

It is desirable that the license names in the `greenLicenses` section be
valid license IDs defined in https://spdx.org/licenses/ whenever possible.

## Interface as a Library

You can also use `js-green-licenses` as a library as well as a command-line
utility. Usually the `LicenseChecker` class is the only one you would have to
use.

### Instantiation

```javascript
const opts = {
  dev: false,
  verbose: true,
};
const checker = new LicenseChecker(opts);
```

Both the `dev` and the `verbose` fields are optional and default to false.
When `dev` is true, the `devDependencies` section is checked as well as the
`dependencies` section of `package.json`. When `verbose` is true, `jsgl`
generates more verbose output.

### Use in Gulp

```javascript
const jsgl = require('js-green-licenses');

gulp.task('check_licenses', function() {
  const checker = new jsgl.LicenseChecker({
    dev: true,
    verbose: false,
  });
  checker.setDefaultHandlers();
  return checker.checkLocalDirectory('.');
});
```

### Methods

*   `LicenseChecker#setDefaultHandler()`

    ```typescript
    setDefaultHandlers(): void;
    ```

    Sets the default event handlers that are used by the CLI. For events
    emitted by `LicenseChecker`, see the [*Events*](#hd-events) subsection.

*   `LicenseChecker#checkLocalDirectory()`

    ```typescript
    checkLocalDirectory(directory: string): Promise<void>;
    ```

    This provides the functionality of the CLI when the `--local` flag is
    passed. It finds and checks the `package.json` file in the `directory` and
    recursively checks its dependencies. This method also detects monorepos
    and checks sub-packages as well, as explained in the [CLI section](#cli)
    above.

    This method reads in the configuration from the `js-green-licenses.json`
    file in the `directory`, if it exists.

*   `LicenseChecker#checkRemotePackage()`

    ```typescript
    checkRemotePackage(pkg: string): Promise<void>;
    ```

    This provides the functionality of the CLI when neither `--local` or
    `--pr` is passed. It retrieves and checks the `package.json` for the
    remote npm package and recursively checks its dependencies.

    This method reads in the configuration from the `js-green-licenses.json`
    file in the current directory of the Node.js process.

*   `LicenseChecker#checkGitHubPR()`

    ```typescript
    checkGitHubPR(repo: GitHubRepository, mergeCommitSha): Promise<void>;
    ```

    This provides the functionality of the CLI when the `--pr` flag is
    passed. It retrieves the `package.json` file from the GitHub repository
    at the given commit SHA and checks its license and recursively checks its
    dependencies. This method also detects monorepos and checks sub-packages
    as well, as explained in the [CLI section](#cli) above.

    This method reads in the configuration from the `js-green-licenses.json`
    file in the repository, if it exists.

    `GitHubRepository` is a helper class for interacting with the GitHub API.
    You can create its instance by calling
    `LicenseChecker#prPathToGitHubRepoAndId()`.

*   `LicenseChecker#prPathToGitHubRepoAndId()`

    ```typescript
    prPathToGitHubRepoAndId(prPath: string): {
        repo: GitHubRepository;
        prId: string;
    };
    ```

    `prPath` must be in the form, `<owner>/<repo>/pull/<id>`. This method
    will return the `GitHubRepository` instance and the PR id for the
    `prPath`.

### <a name="hd-events"></a>Events

A `LicenseChecker` object emits following events during its processing.

*   `non-green-license`
    Emitted when a package with a non-green license is detected. The argument is
    ```typescript
    interface NonGreenLicense {
      packageName: string;
      version: string;
      licenseName: string|null;
      parentPackages: string[];
    }
    ```

*   `package.json`
    Emitted for each `package.json` file being checked. This is emitted only
    when checking local repositories or GitHub repositories, but not when
    checking remote packages.

    The argument is a file path string of the corresponding `package.json` file.

*   `end`
    Emitted when the processing is done. No argument is given.

*   `error`
    Emitted when an error occurrs while processing. The argument is
    ```typescript
    interface CheckError {
      err: Error;
      packageName: string;
      versionSpec: string;
      parentPackages: string[];
    }
    ```

[actions-image]: https://github.com/google/js-green-licenses/workflows/ci/badge.svg
[actions-url]: https://github.com/google/js-green-licenses/actions
[codecov-url]: https://codecov.io/gh/google/js-green-licenses
[david-image]: https://david-dm.org/google/js-green-licenses.svg
[david-url]: https://david-dm.org/google/js-green-licenses
[gts-image]: https://img.shields.io/badge/code%20style-google-blueviolet.svg
[gts-url]: https://github.com/google/gts
[npm-image]: https://img.shields.io/npm/v/js-green-licenses.svg
[npm-url]: https://npmjs.org/package/js-green-licenses
[snyk-image]: https://snyk.io/test/github/google/js-green-licenses/badge.svg
[snyk-url]: https://snyk.io/test/github/google/js-green-licenses


CHANGELOG.md
# Changelog

[npm history][1]

[1]: https://www.npmjs.com/package/js-green-licenses?activeTab=versions

## [4.0.0](https://github.com/google/js-green-licenses/compare/v3.0.1...v4.0.0) (2023-01-09)


### ⚠ BREAKING CHANGES

* drop support for node.js 10.x ([#176](https://github.com/google/js-green-licenses/issues/176))

### Bug Fixes

* **deps:** update dependency gaxios to v5 ([#195](https://github.com/google/js-green-licenses/issues/195)) ([876307f](https://github.com/google/js-green-licenses/commit/876307f4c9fd70eebb56f32f1f19a3ae2500b462))
* **deps:** update dependency package-json to v7 ([#172](https://github.com/google/js-green-licenses/issues/172)) ([cb2900e](https://github.com/google/js-green-licenses/commit/cb2900e8dd9040ba8fc4f694b407164c2f7d827b))
* **deps:** upgrade to typescript 4.4.4 ([#180](https://github.com/google/js-green-licenses/issues/180)) ([da3fed5](https://github.com/google/js-green-licenses/commit/da3fed520287b6ca8e82426f8d95811baf876e64))


### Build System

* drop support for node.js 10.x ([#176](https://github.com/google/js-green-licenses/issues/176)) ([786b01a](https://github.com/google/js-green-licenses/commit/786b01a8fecb08ce275c853b34cb946a9fa0f3cd))

## [4.0.0](https://www.github.com/google/js-green-licenses/compare/v3.0.1...v4.0.0) (2021-10-20)


### ⚠ BREAKING CHANGES

* drop support for node.js 10.x (#176)

### Bug Fixes

* **deps:** update dependency package-json to v7 ([#172](https://www.github.com/google/js-green-licenses/issues/172)) ([cb2900e](https://www.github.com/google/js-green-licenses/commit/cb2900e8dd9040ba8fc4f694b407164c2f7d827b))
* **deps:** upgrade to typescript 4.4.4 ([#180](https://www.github.com/google/js-green-licenses/issues/180)) ([da3fed5](https://www.github.com/google/js-green-licenses/commit/da3fed520287b6ca8e82426f8d95811baf876e64))


### Build System

* drop support for node.js 10.x ([#176](https://www.github.com/google/js-green-licenses/issues/176)) ([786b01a](https://www.github.com/google/js-green-licenses/commit/786b01a8fecb08ce275c853b34cb946a9fa0f3cd))

### [3.0.1](https://www.github.com/google/js-green-licenses/compare/v3.0.0...v3.0.1) (2021-04-05)


### Bug Fixes

* handle incorrect licenses fields ([#146](https://www.github.com/google/js-green-licenses/issues/146)) ([209847f](https://www.github.com/google/js-green-licenses/commit/209847fa96d3f8676cfb292ef879a7e58e4cdecd))

## [3.0.0](https://www.github.com/google/js-green-licenses/compare/v2.0.1...v3.0.0) (2021-01-22)


### ⚠ BREAKING CHANGES

* The `packageWhitelist` field has been renamed to `packageAllowlist`.

### Features

* added skipping of package.json check if package is whitelisted (resolves [#118](https://www.github.com/google/js-green-licenses/issues/118)) ([#136](https://www.github.com/google/js-green-licenses/issues/136)) ([0d95d18](https://www.github.com/google/js-green-licenses/commit/0d95d18eb146bb8a003e741789729cf88d7e759c))
* use allowlist for allowed licenses ([#153](https://www.github.com/google/js-green-licenses/issues/153)) ([d24827f](https://www.github.com/google/js-green-licenses/commit/d24827fa25d66020f876412f706a9dfbd8b56a98))


### Bug Fixes

* **deps:** update dependency gaxios to v4 ([#148](https://www.github.com/google/js-green-licenses/issues/148)) ([9b66f9c](https://www.github.com/google/js-green-licenses/commit/9b66f9cd1768c4a7f5dde454192a9dc0f0d10c8f))
* **deps:** use meow for argument parsing ([#156](https://www.github.com/google/js-green-licenses/issues/156)) ([ae07bc8](https://www.github.com/google/js-green-licenses/commit/ae07bc81f21a7be753424c4dfc3fed4c6ee69060))

### [2.0.1](https://www.github.com/google/js-green-licenses/compare/v2.0.0...v2.0.1) (2020-05-18)


### Bug Fixes

* Added support for the UNLICENSED NPM magic value (fixes [#113](https://www.github.com/google/js-green-licenses/issues/113)) ([#134](https://www.github.com/google/js-green-licenses/issues/134)) ([c7966eb](https://www.github.com/google/js-green-licenses/commit/c7966eb39a95c3376f3845f7797bb3d72f2c5904))

## [2.0.0](https://www.github.com/google/js-green-licenses/compare/v1.1.0...v2.0.0) (2020-05-07)


### ⚠ BREAKING CHANGES

* drop support for node.js 8 (#124)

### Bug Fixes

* **deps:** update dependency gaxios to v3 ([#121](https://www.github.com/google/js-green-licenses/issues/121)) ([1fdd3e4](https://www.github.com/google/js-green-licenses/commit/1fdd3e4e7ad642da55ad94b72d71049f4789310c))
* **deps:** update dependency npm-package-arg to v7 ([#110](https://www.github.com/google/js-green-licenses/issues/110)) ([25d0235](https://www.github.com/google/js-green-licenses/commit/25d0235d0ff996b644fe84482241bda5a4b14b2a))
* **deps:** update dependency npm-package-arg to v8 ([#111](https://www.github.com/google/js-green-licenses/issues/111)) ([4763858](https://www.github.com/google/js-green-licenses/commit/4763858fe10d35e2f1e944d0f4e91a309503c812))
* add 0BSD as a valid license ([#132](https://www.github.com/google/js-green-licenses/issues/132)) ([e4db3eb](https://www.github.com/google/js-green-licenses/commit/e4db3ebe01bf49a922f0d3a0d71d9ee4af83cdf0))


### Build System

* drop support for node.js 8 ([#124](https://www.github.com/google/js-green-licenses/issues/124)) ([06283e1](https://www.github.com/google/js-green-licenses/commit/06283e10c44306f88cdb382ac9d593f4a48c0b7d))

## [1.1.0](https://www.github.com/google/js-green-licenses/compare/v1.0.0...v1.1.0) (2019-06-05)


### Features

* support local paths ([#101](https://www.github.com/google/js-green-licenses/issues/101)) ([6d08407](https://www.github.com/google/js-green-licenses/commit/6d08407))

## [1.0.0](https://www.github.com/google/js-green-licenses/compare/v0.5.0...v1.0.0) (2019-05-29)


### ⚠ BREAKING CHANGES

* Drop support for Node.js 6.x which reaching EOL in a
couple of weeks.

### Bug Fixes

* fails to check monorepo with non-published packages ([#89](https://www.github.com/google/js-green-licenses/issues/89)) ([96a7279](https://www.github.com/google/js-green-licenses/commit/96a7279))

src/checker.ts
// Copyright 2017 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {EventEmitter} from 'events';
import * as fs from 'fs';
import npmPackageArg from 'npm-package-arg';
import packageJson from 'package-json';
import * as path from 'path';
import spdxCorrect from 'spdx-correct';
import spdxSatisfies from 'spdx-satisfies';
import {inspect, promisify} from 'util';
import semver from 'semver';

import * as config from './config';
import {GitHubRepository} from './github';
import {
  Dependencies,
  ensurePackageJson,
  PackageJson,
} from './package-json-file';

export {GitHubRepository} from './github';

const fsAccess = promisify(fs.access);
const fsReadDir = promisify(fs.readdir);
const fsReadFile = promisify(fs.readFile);

// Valid license IDs defined in https://spdx.org/licenses/ must be used whenever
// possible. When adding new licenses, please consult the relevant documents.
const DEFAULT_GREEN_LICENSES = [
  '0BSD',
  'AFL-2.1',
  'AFL-3.0',
  'APSL-2.0',
  'Apache-1.1',
  'Apache-2.0',
  'Artistic-1.0',
  'Artistic-2.0',
  'BSD-2-Clause',
  'BSD-3-Clause',
  'BSL-1.0',
  'CC-BY-1.0',
  'CC-BY-2.0',
  'CC-BY-2.5',
  'CC-BY-3.0',
  'CC-BY-4.0',
  'CC0-1.0',
  'CDDL-1.0',
  'CDDL-1.1',
  'CPL-1.0',
  'EPL-1.0',
  'FTL',
  'IPL-1.0',
  'ISC',
  'LGPL-2.0',
  'LGPL-2.1',
  'LGPL-3.0',
  'LPL-1.02',
  'MIT',
  'MPL-1.0',
  'MPL-1.1',
  'MPL-2.0',
  'MS-PL',
  'NCSA',
  'OpenSSL',
  'PHP-3.0',
  'Ruby',
  'Unlicense',
  'W3C',
  'Xnet',
  'ZPL-2.0',
  'Zend-2.0',
  'Zlib',
  'libtiff',
];

// options for constructing LicenseChecker
export interface LicenseCheckerOptions {
  dev?: boolean; // also check devDependencies
  verbose?: boolean; // verbose error outputs
}

// argument for 'non-green-license' event from LicenseChecker
export interface NonGreenLicense {
  packageName: string;
  version: string;
  licenseName: string | null;
  parentPackages: string[];
}

// argument for 'error' event from LicenseChecker
export interface CheckError {
  err: Error;
  packageName: string;
  versionSpec: string;
  parentPackages: string[];
}

export interface DefaultHandlerOptions {
  setExitCode?: boolean;
}

type EventType = 'non-green-license' | 'package.json' | 'end' | 'error';

export class LicenseChecker extends EventEmitter {
  // Cache for packageName@version's that are already processed.
  private readonly processedPackages: Set<string> = new Set();
  // Cache for packageName@version's that failed for fetching.
  private readonly failedPackages: Set<string> = new Set();
  // Local packages, for monorepo
  private readonly localPackages: Set<string> = new Set<string>();
  private readonly opts: LicenseCheckerOptions;
  private config: config.Config = {};
  // Licenses in this expression must be valid license IDs defined in
  // https://spdx.org/licenses/.
  private greenLicenseExpr = '';
  // List of license names that are not SPDX-conforming IDs but are allowed.
  private allowlistedLicenses: string[] = [];

  constructor({dev = false, verbose = false}: LicenseCheckerOptions = {}) {
    super();
    this.opts = {dev, verbose};
  }

  on(
    event: 'non-green-license',
    listener: (arg: NonGreenLicense) => void
  ): this;
  // 'package.json' events are not emitted for remote packages.
  on(event: 'package.json', listener: (filePath: string) => void): this;
  on(event: 'end', listener: () => void): this;
  on(event: 'error', listener: (checkError: CheckError) => void): this;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  on(event: EventType, listener: (...args: any[]) => void): this {
    return super.on(event, listener);
  }

  emit(event: 'non-green-license', arg: NonGreenLicense): boolean;
  emit(event: 'package.json', filePath: string): boolean;
  emit(event: 'end'): boolean;
  emit(event: 'error', checkError: CheckError): boolean;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  emit(event: EventType, ...args: any[]): boolean {
    return super.emit(event, ...args);
  }

  private init(cfg: config.Config | null): void {
    this.config = cfg || {};

    const greenLicenses = this.config.greenLicenses || DEFAULT_GREEN_LICENSES;
    const validGreenLicenses: string[] = [];
    const invalidGreenLicenses: string[] = [];
    for (const license of greenLicenses) {
      const corrected = this.correctLicenseName(license);
      if (corrected) {
        validGreenLicenses.push(corrected);
      } else {
        invalidGreenLicenses.push(license);
      }
    }
    this.greenLicenseExpr = `(${validGreenLicenses.join(' OR ')})`;
    this.allowlistedLicenses = invalidGreenLicenses;

    this.processedPackages.clear();
    this.failedPackages.clear();
  }

  private getLicense(pkgJson: Partial<PackageJson>): string | null {
    // Some package.json files have incorrect license fields, and old packages
    // may have legacy licence field format. See
    // https://docs.npmjs.com/files/package.json#license for details. The code
    // below is a little complicated to deal with those cases.
    const license = pkgJson.license || pkgJson.licenses;
    if (!license) {
      if (pkgJson.private) {
        return 'private';
      }
      return null;
    }
    if (typeof license === 'string') {
      return license;
    }
    if (Array.isArray(license)) {
      if (license.length === 0) {
        return null;
      }
      const types = license.map(x => x.type).filter(x => !!x);
      return types.length === 1 ? types[0] : `(${types.join(' OR ')})`;
    }
    return license.type || null;
  }

  private correctLicenseName(license: string): string | null {
    // NPM specific value.
    if (license === 'UNLICENSED' || license === 'UNLICENCED') {
      console.warn(`Unlicensed package, specified license: ${license}`);
      return 'UNLICENSED';
    }

    const corrected = spdxCorrect(license);
    if (this.opts.verbose && corrected && corrected !== license) {
      console.warn(`Correcting ${license} to ${corrected}`);
    }
    return corrected;
  }

  private isPackageAllowlisted(packageName: string): boolean {
    return (
      !!this.config.packageAllowlist &&
      this.config.packageAllowlist.some(p => p === packageName)
    );
  }

  private isGreenLicense(license: string | null): boolean {
    if (!license) return false;

    const correctedName = this.correctLicenseName(license);
    // `license` is not a valid or correctable SPDX id. Check the allowlist.
    if (!correctedName) {
      return this.allowlistedLicenses.some(l => l === license);
    }

    try {
      return spdxSatisfies(correctedName, this.greenLicenseExpr);
    } catch (e) {
      const err = e as Error;
      // Most likely because license is not recognized. Just return false.
      if (this.opts.verbose) {
        console.error(err.message || err);
      }
      return false;
    }
  }

  private async getPackageJson(
    packageName: string,
    versionSpec: string,
    localDirectory: string | null
  ): Promise<{}> {
    // If this has a relative URL, and is a local package, find the package json from the
    // indicated directory
    if (versionSpec.startsWith('file:') && localDirectory) {
      const relativePath = versionSpec.slice('file:'.length);
      const packageJsonPath = path.join(
        localDirectory,
        relativePath,
        'package.json'
      );
      this.emit('package.json', packageJsonPath);
      const contents = await fsReadFile(packageJsonPath, 'utf8');
      return JSON.parse(contents);
    }
    return packageJson(packageName, {
      version: versionSpec,
      fullMetadata: true,
    });
  }

  private async checkLicenses(
    packageName: string,
    versionSpec: string,
    localDirectory: string | null,
    ...parents: string[]
  ): Promise<void> {
    const spec = `${packageName}@${versionSpec}`;
    if (this.failedPackages.has(spec)) return;

    // remove tilde/caret to check for an exact version, ^0.5.0-rc.0 becomes 0.5.0-rc.0
    const version = versionSpec.replace(/^[^~]/, '');
    // if the dependency is a local package then skip verification at this step. will be checked independently
    if (this.localPackages.has(`${packageName}@${version}`)) return;

    try {
      const json = await this.getPackageJson(
        packageName,
        versionSpec,
        localDirectory
      );
      await this.checkPackageJson(
        json,
        packageName,
        localDirectory,
        ...parents
      );
    } catch (e) {
      const err = e as Error;
      this.failedPackages.add(spec);
      this.emit('error', {
        err,
        packageName,
        versionSpec,
        parentPackages: parents,
      });
    }
  }

  private async checkLicensesForDeps(
    deps: Dependencies | undefined,
    localDirectory: string | null,
    ...parents: string[]
  ): Promise<void> {
    if (!deps) return;
    for (const pkg of Object.keys(deps)) {
      const depVersion = deps[pkg];
      await this.checkLicenses(pkg, depVersion, localDirectory, ...parents);
    }
  }

  private async checkPackageJson(
    json: Partial<PackageJson>,
    packageName: string | null,
    localDirectory: string | null,
    ...parents: string[]
  ): Promise<void> {
    packageName = (packageName || json.name || 'undefined') as string;

    const isAllowlisted = this.isPackageAllowlisted(packageName);
    if (isAllowlisted) {
      json.version = semver.valid(json.version) ? json.version : '0.0.0';
    } else {
      ensurePackageJson(json);
    }

    if (json.name !== packageName) {
      console.warn(
        `Package name mismatch. Expected ${packageName}, but got ${json.name}`
      );
    }

    const pkgVersion = json.version;
    const packageAndVersion = `${packageName}@${pkgVersion}`;
    if (this.processedPackages.has(packageAndVersion)) return;
    this.processedPackages.add(packageAndVersion);

    if (this.isPackageAllowlisted(packageName)) {
      console.log(`${packageName} is allowlisted.`);
    } else {
      const license = this.getLicense(json);
      if (!this.isGreenLicense(license)) {
        this.emit('non-green-license', {
          packageName,
          version: pkgVersion || 'undefined',
          licenseName: license,
          parentPackages: parents,
        });
      }
    }

    await this.checkLicensesForDeps(
      json.dependencies,
      localDirectory,
      ...parents,
      packageAndVersion
    );
    if (this.opts.dev) {
      await this.checkLicensesForDeps(
        json.devDependencies,
        localDirectory,
        ...parents,
        packageAndVersion
      );
    }
  }

  private async checkPackageJsonContent(
    content: string,
    localDirectory: string | null
  ): Promise<void> {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    let json: any = null;
    try {
      json = JSON.parse(content);
      await this.checkPackageJson(json, json.name, localDirectory);
    } catch (e) {
      const err = e as Error;
      const packageName = json?.name || '(unknown package)';
      const versionSpec = json?.version || '(unknown version)';
      this.emit('error', {
        err,
        packageName,
        versionSpec,
        parentPackages: [],
      });
    }
  }

  private async getLocalPackageJsonFiles(directory: string): Promise<string[]> {
    const packageJsons: string[] = [];
    const addPackageJson = async (dir: string) => {
      try {
        const pj = path.join(dir, 'package.json');
        await fsAccess(pj);
        packageJsons.push(pj);
      } catch {
        // package.json doesn't exist. Ignore.
      }
    };

    // Find the top-level package.json first.
    await addPackageJson(directory);

    // Find `packages/<name>/package.json` files in case this is a monorepo.
    try {
      const packages = path.join(directory, 'packages');
      const subpackages = await fsReadDir(packages);
      // This is a monorepo. Find package.json from each directory under
      // `packages`.
      for (const dir of subpackages) {
        await addPackageJson(path.join(packages, dir));
      }
    } catch {
      // The `packages` directory doesn't exist. Not a monorepo. Return just the
      // top-level package.json.
    }

    return packageJsons;
  }

  async checkLocalDirectory(directory: string): Promise<void> {
    this.init(await config.getLocalConfig(directory));
    const packageJsons = await this.getLocalPackageJsonFiles(directory);
    if (packageJsons.length === 0) {
      console.log('No package.json files have been found.');
    }
    for (const pj of packageJsons) {
      const content = await fsReadFile(pj, 'utf8');
      const json = JSON.parse(content);
      if (json && json.name && json.version) {
        this.localPackages.add(`${json.name}@${json.version}`);
      }
    }
    for (const pj of packageJsons) {
      this.emit('package.json', pj);
      const content = await fsReadFile(pj, 'utf8');
      await this.checkPackageJsonContent(content, path.dirname(pj));
    }
    this.emit('end');
  }

  async checkRemotePackage(pkg: string): Promise<void> {
    // For checking remote packages, use config file in the current directory.
    this.init(await config.getLocalConfig(process.cwd()));
    const pkgArgs = npmPackageArg(pkg);
    const pkgType = pkgArgs.type;
    if (!['tag', 'version', 'range'].some(type => type === pkgType)) {
      throw new Error(`Unsupported package spec: ${pkg}`);
    }
    if (!pkgArgs.name || !pkgArgs.fetchSpec) {
      throw new Error(`Invalid package spec: ${pkg}`);
    }
    await this.checkLicenses(pkgArgs.name, pkgArgs.fetchSpec, null);
    this.emit('end');
  }

  /** @param prPath Must be in a form of <owner>/<repo>/pull/<id>. */
  prPathToGitHubRepoAndId(prPath: string): {
    repo: GitHubRepository;
    prId: number;
  } {
    const regexp = /^([^/]+)\/([^/]+)\/pull\/(\d+)$/;
    const matched = regexp.exec(prPath);
    if (!matched) {
      throw new Error(
        `Invalid github pull request path: ${prPath}. ` +
          'Must be in the form <owner>/<repo>/pull/<id>.'
      );
    }
    const [, owner, repoName, prId] = matched;
    return {repo: new GitHubRepository(owner, repoName), prId: Number(prId)};
  }

  async checkGitHubPR(
    repo: GitHubRepository,
    mergeCommitSha: string
  ): Promise<void> {
    this.init(await config.getGitHubConfig(repo, mergeCommitSha));
    const packageJsons = await repo.getPackageJsonFiles(mergeCommitSha);
    if (packageJsons.length === 0) {
      console.log('No package.json files have been found.');
    }
    for (const pj of packageJsons) {
      this.emit('package.json', pj.filePath);
      await this.checkPackageJsonContent(pj.content, null);
    }
    this.emit('end');
  }

  /** set default event handlers for CLI output. */
  setDefaultHandlers(options: DefaultHandlerOptions = {}): void {
    let nonGreenCount = 0;
    let errorCount = 0;
    this.on(
      'non-green-license',
      ({packageName, version, licenseName, parentPackages}) => {
        nonGreenCount++;
        const licenseDisplay = licenseName || '(no license)';
        const packageAndVersion = `${packageName}@${version}`;
        console.log(`${licenseDisplay}: ${packageAndVersion}`);
        console.log(`  ${[...parentPackages, packageAndVersion].join(' -> ')}`);
        console.log();
      }
    )
      .on('package.json', filePath => {
        console.log(`Checking ${filePath}...`);
        console.log();
      })
      .on('error', ({err, packageName, versionSpec, parentPackages}) => {
        errorCount++;
        const packageAndVersion = `${packageName}@${versionSpec}`;
        console.log(`Error while checking ${packageAndVersion}:`);
        console.log(`  ${[...parentPackages, packageAndVersion].join(' -> ')}`);
        console.log();
        console.log(`${inspect(err)}`);
        console.log();
      })
      .on('end', () => {
        if (nonGreenCount > 0 || errorCount > 0) {
          if (options.setExitCode) {
            process.exitCode = 1;
          }
          if (nonGreenCount > 0) {
            console.log(`${nonGreenCount} non-green licenses found.`);
          }
          if (errorCount > 0) {
            console.log(`${errorCount} errors found.`);
          }
        } else {
          console.log('All green!');
        }
      });
  }
}
